File: binary_search_tree.py
class TreeNode():
  def __init__(self, data):
    self.data = data
    self.left = None
    self.right = None

class BinarySearchTree():
  def __init__(self, root=None):
    self.root = None

  def insert(self, data, node=None):

    # default case, if no node, start at root
    if not node:
      node = self.root
    
    # no root
    if not self.root:
      self.root = TreeNode(data)
      return self.root
    else:
      # there is a root
      if data < node.data and node.left == None:
        # the data is less than the curr node, and curr node has no left
        # initaite a new node as the curr node's left child
        node.left = TreeNode(data)
      elif data < node.data:
        # there is a left node, so recurse down the left side, passing in the left node to the insert function
        node.left = self.insert(data, node.left)
      elif data > node.data and node.right == None:
        # if the data is greater than curr node, and there is no right child
        # initiate a new node as the curr node's right child
        node.right = TreeNode(data)
      elif data > node.data:
        # the is a right node, so recurse down the right side, passing in the right node to the insert funciton
        node.right = self.insert(data, node.right)
    return node

  # depth-first-search
  def in_order(self, node=None):
    #left, root, right
    if not node:
      node = self.root
    if node:
      if node.left:
        self.in_order(node.left)
      print(node.data)

      if node.right:
        self.in_order(node.right)

  def pre_order(self, node=None):
    # root, left right
    if not node:
      node = self.root
    if node:
      print(node.data)
      if node.left:
        self.pre_order(node.left)
      if node.right:
        self.pre_order(node.right)
  
  def post_order(self, node=None):
    # left, right, root
    if not node:
      node = self.root
    if node:
      if node.left:
        self.post_order(node.left)
      if node.right:
        self.post_order(node.right)
      print(node.data)

bst = BinarySearchTree()
bst.insert(5)
bst.insert(10)
bst.insert(1)
bst.insert(14)
bst.insert(16)
bst.insert(19)
bst.insert(50)
bst.insert(20)
bst.insert(3)
bst.insert(7)
print('  ', bst.root.data)
print(' ', bst.root.left.data, ' ',  bst.root.right.data)
bst.in_order()

File: binarySearchTree.js
class TreeNode {
  constructor(data) {
    this.data = data;
    this.left_child = null;
    this.right_child = null;
  }
}

class BinarySearchTree {
  constructor() {
    this.root = null;
  }

  insert(data, node=null) {
    // if starting node wasn't passed in, start at the root (usually the case, only not the case when called recursively)
    if (!node) {
      var node = this.root;
    }
    // if no root, make the inserted node the root
    if (!this.root) {
      this.root = new TreeNode(data);
      return this.root;
    } else {
      // there is a root
      if (data < node.data && node.left_child === null) {
        // belongs on the left and there is no left child
        node.left_child = new TreeNode(data);
      } else if (data < node.data) {
        // new node belongs on the left, and there is a left_child
        node.left_child = this.insert(data, node.left_child);
      } else if (data > node.data && node.right_child === null) {
        // new node belongs on the right and there is no right_child
        node.right_child = new TreeNode(data)
      } else if (data > node.data) {
        // new node belongs on the right and there is a right_child
        node.right_child = this.insert(data, node.right_child);
      }
    }
    return node;
  }

  inOrder(node=null) {
    // left, root, right
    if (!node) {
      node = this.root;
    }
    if (node) {
      if(node.left_child) {
        this.inOrder(node.left_child);
      }
      console.log(node.data);

      if (node.right_child) {
        this.inOrder(node.right_child);
      }
    }
  }

  preOrder(node=null) {
    // root, left, right
    if (!node) {
      node = this.root;
    }
    if (node) {
      console.log(node.data);
      if (node.left_child) {
        this.preOrder(node.left_child)
      }
      if (node.right_child) {
        this.preOrder(node.right_child)
      }
    }
  }

  postOrder(node=null) {
    // left, right, root
    if (!node) {
      node = this.root;
    }
    if (node) {
      if (node.left_child) {
        this.postOrder(node.left_child);
      }
      if (node.right_child) {
        this.postOrder(node.right_child);
      }
      console.log(node.data)
    }
  }
}

const tree = new BinarySearchTree()
tree.insert(8)
tree.insert(4)
tree.insert(16)
tree.insert(5)
tree.insert(2)
console.log('left, root, right')
console.log(tree.inOrder())

console.log('root left right')
console.log(tree.preOrder())

console.log('left, right, root')
console.log(tree.postOrder())